pkg/sources/circleci/circleci.go

```diff
@@ -129,115 +170,147 @@ func (s *Source) Chunks(ctx context.Context, chunksChan chan *sources.Chunk, _ .
 	return nil
 }
 
-type project struct {
-	VCS      string `json:"vcs_type"`
-	Username string `json:"username"`
-	RepoName string `json:"reponame"`
-}
-
-func (s *Source) projects(_ context.Context) ([]project, error) {
-	reqURL := fmt.Sprintf("%sprojects", baseURL)
-	req, err := http.NewRequest("GET", reqURL, nil)
+// listAllProjects lists all the projects the token can access
+func (s *Source) listAllProjects(ctx context.Context) ([]project, error) {
+	ctx.Logger().V(5).Info("listing projects")
+	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("%s/projects", baseURL), http.NoBody)
 	if err != nil {
 		return nil, err
 	}
+
 	req.Header.Set("Circle-Token", s.token)
 	req.Header.Set("Accept", "application/json")
-	res, err := s.client.Do(req)
+
+	resp, err := s.client.Do(req)
 	if err != nil {
 		return nil, err
 	}
-	defer res.Body.Close()
+	defer func() {
+		_, _ = io.Copy(io.Discard, resp.Body)
+		_ = resp.Body.Close()
+	}()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		var projects []project
+		if err := json.NewDecoder(resp.Body).Decode(&projects); err != nil {
+			return nil, fmt.Errorf("cannot decode CircleCI projects JSON: %w", err)
+		}
 
-	if res.StatusCode > 399 && res.StatusCode < 500 {
-		return nil, fmt.Errorf("invalid credentials, status %d", res.StatusCode)
+		ctx.Logger().V(2).Info(fmt.Sprintf("successfully listed %d projects", len(projects)))
```

Comment:
1. Also more great logging
2. It's fine to leave as-is, but I find whenever I'm using `fmt.Sprintf` in a logging call I probably ought to just make it a param, e.g:

```go
ctx.Logger().V(2).Info("successfully listed projects", "project_count", len(projects))
```

The main thing here is it makes searching through logs easier--you can just do `= "successfully listed projects"` instead of `=~ "successfully listed \d+ projects"`